core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java

```diff
@@ -136,7 +136,7 @@
     /**
      * An instance map for reusing converters, exception handlers and decorators.
      */
-    private static final ConcurrentMap<Class<?>, Object> instanceCache = new ConcurrentHashMap<>();
+    private static final ConcurrentMap<Class<?>, Object> instanceCache = new MapMaker().weakKeys().makeMap();
```

Comment:
By the way maybe for a separate PR but I recently learned of `ClassValue`

https://docs.oracle.com/javase/7/docs/api/java/lang/ClassValue.html

It's like a map from class to value, but handled natively by Java with optimal performance.

Reply:
Thanks! I have not known that class. 😄

Reply:
Here's an example :) https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/c6eee706604a9edf35758d90dc974e4ee82a0f0a/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/decorator/BaseDecorator.java#L25

Reply:
Fixed. 😉